// helm-analytics/index.js
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));

class HelmAnalytics {
  constructor(options = {}) {
    this.siteId = options.siteId || process.env.HELM_SITE_ID;
    this.apiUrl = options.apiUrl || 'https://api-sentinel.getmusterup.com/track';
    
    if (!this.siteId) {
      console.warn('HelmAnalytics: No Site ID provided. Tracking will be disabled.');
    }
  }

  // Generic track method
  async track(req, eventType = 'pageview', metadata = {}) {
    if (!this.siteId) return;

    try {
      const payload = {
        siteId: this.siteId,
        url: req.originalUrl || req.url,
        userAgent: req.headers['user-agent'] || '',
        referrer: req.headers['referer'] || req.headers['referrer'] || '',
        // ClientIP logic: Headers or socket
        clientIp: req.headers['x-forwarded-for'] ? req.headers['x-forwarded-for'].split(',')[0].trim() : (req.socket ? req.socket.remoteAddress : ''),
        eventType: eventType,
        isServerSide: true,
        ...metadata
      };

      // Fire and forget (don't await in middleware usually, but here we define the promise)
      fetch(this.apiUrl, {
        method: 'POST',
        body: JSON.stringify(payload),
        headers: { 'Content-Type': 'application/json' }
      }).catch(err => {
        // Silent error in background
      });

    } catch (err) {
      // Catch synchronous errors
    }
  }

  // Express Middleware
  middleware() {
    return (req, res, next) => {
      // don't track static files if easy to detect, but usually backend handles API routes
      this.track(req);
      next();
    };
  }
}

module.exports = HelmAnalytics;
